[RFC] replace sfdisk w/ python library in rouge

sa-kib opened the pull request:
`sfdisk` sucks, can't create GPT images for UFS device. Aim for more flexibility with native gpt-image library.

lorc reviewed (changes requested):
I am totally happy with the general approach, but there are couple of high-level comments:

1. We'd better include this library as external dependency, no need to include it into `moulin` source base.
2. `Disk` module from the lib does not suit us for couple of reasons:
a. It does not like existing files
b. It fills the image file with zeroes, which is very bad for storage. Right now `rouge` produces almost empty sparse images, the GPT lib will break this feature
3. You didn't covered case when we have GPT in GPT (like Android GPT table/partition in our full image). 
4. Taking 2 and 3 into account, I propose to not use `Disk` module, but use low-level functionality like `Table` and `Geometry`. Yes, this will require more code, but we will have better control on the resulting image.
5. You can remove old `sfdisk.py` at all.

sa-kib commented on moulin/rouge/gpt_image/disk.py:
```diff
@@ -1,3 +1,5 @@
+# SPDX-License-Identifier: MIT
```
dependency introduced in [3be78aa](https://github.com/xen-troops/moulin/pull/73/commits/3be78aa4aeab2c211d6f484b518c5e39f5eaec9a)

sa-kib commented on moulin/rouge/gpt_image/disk.py:
```diff
@@ -0,0 +1,151 @@
+        self.table = Table(self.geometry)
+        # @TODO: move this into the write method
+        with open(self.image_path, "r+b") as f:
+            # zero entire disk
```
done in [ad4425b](https://github.com/xen-troops/moulin/pull/73/commits/ad4425b48328b8fb2db07c0c29f4327e233a76d0)

sa-kib commented on docs/rouge.rst:
```diff
@@ -154,6 +154,10 @@ If actual size is bigger than specified - error will be printed with
 
+The sector size of resulting disk image is 512 bytes by default,
+which corresponds to most block storage devices found in embedded devices.
+Custom sector size can be provided via :code:`sector_size` property.
```
yes, fixed in [432e8d3](https://github.com/xen-troops/moulin/pull/73/commits/432e8d3f889a9e0fee38f17e7aa59ceaf469ec1a)

sa-kib commented on moulin/rouge/block_entry.py:
```diff
@@ -9,12 +9,14 @@
 import logging
 import itertools
-from typing import List, Tuple, NamedTuple, cast
+from typing import List, Tuple, NamedTuple, cast, Any
```
nowhere really, dropped

sa-kib commented on moulin/rouge/block_entry.py:
```diff
@@ -89,13 +95,23 @@ def _process_entry(node: YamlValue):
             self._complete_init()
         if offset == 0:
-            sfdisk.write(fp, self._partitions)
+            DEFAULT_ALIGNMENT = 1 * 1024 * 1024  # 1 MiB
```
this constant is move to wrapper now, just as in `sfdisk.py`

lorc commented on moulin/rouge/gpti.py:
```diff
@@ -0,0 +1,81 @@
+    # Account for GPT copy
+    return ret, end + 16 * 1024 * 1024
+
+def write(fileo: BinaryIO, partitions: List[Any], size: int, sector_size: int):
```
It will be good to add `offset` parameter...

lorc commented on moulin/rouge/gpti.py:
```diff
@@ -0,0 +1,81 @@
+            )
+    table.update()
+    with open(image, "r+b") as f:
+        f.seek(0)
```
... and use this `offset` parameter there to support GPT-in-GPT use-case

sa-kib replied:
done

lorc commented on moulin/rouge/gpti.py:
```diff
@@ -0,0 +1,81 @@
+            part.gpt_guid, DEFAULT_ALIGNMENT // sector_size)
+            )
+    table.update()
+    with open(image, "r+b") as f:
```
You already have `fileo` parameter, which is file object to write partition table. No need to open it for a second time.

sa-kib replied:
fixed

lorc commented on moulin/rouge/gpti.py:
```diff
@@ -0,0 +1,81 @@
+
+def write(fileo: BinaryIO, partitions: List[Any], size: int, sector_size: int):
+    image = pathlib.Path(fileo.name)
+    image.touch(exist_ok=True)
```
Do you really need this?

sa-kib replied:
no, removed

lorc commented on moulin/rouge/block_entry.py:
```diff
@@ -50,12 +50,16 @@ class GPT(BlockEntry):
         if _requested_image_size_node:
             self._requested_image_size = _parse_size(_requested_image_size_node)
 
+        if node["sector_size"]:
```
Could you please divide this commit into two? "switch to gpti" and "support custom sector size" are separate changes

sa-kib replied:
yes, done

lorc reviewed (changes requested):
Thanks! I'm generally happy with changes, but there are few minor comments.

Also, looks like CI tools it not happy. You can do local checks with 

`flake8 . --count --max-complexity=12 --max-line-length=120 --statistics`

sa-kib commented:
> Thanks! I'm generally happy with changes, but there are few minor comments.
> 
> Also, looks like CI tools it not happy. You can do local checks with
> 
> `flake8 . --count --max-complexity=12 --max-line-length=120 --statistics`

corrected, also rebased on `main`

lorc reviewed (approved):
Thank you!

`Reviewed-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
